File: compiler/state_discover.py
#!/usr/bin/env python3
"""state_discover.py — find game-state addresses by snapshot diffing.

Workflow:
  1. control.py halt
  2. control.py snapshot                          # S0 = baseline
  3. <perform some action>                        # tap+step, or input the
                                                  # player makes only when a
                                                  # specific variable should
                                                  # change
  4. control.py halt; control.py snapshot         # S1 = after action
  5. state_discover.py diff S0 S1                 # see all bytes that changed

Refinement subcommands:
  changed S0 S1 [S2 S3 ...]
      Addresses where every adjacent pair differs. Use when an action is
      repeated and the variable is expected to change each time (e.g. tick
      counter, walk-right repeated 3x).
  reverted S0 S1 [S2 ...]
      Addresses that differ from S0 in S1 but match S0 again in some later
      Sn — i.e. the action was un-done. The bread-and-butter filter for
      "player_x: walked right then walked back".
  constant S0 S1 [S2 ...]
      Addresses that are identical across every snapshot. Useful as a
      sanity check ("did the snapshots actually come from a paused game?")
      and to subtract from candidate sets.
  monotonic S0 S1 S2 [S3 ...]
      Addresses where the (little-endian, width-typed) value strictly
      increases at every step, OR strictly decreases at every step. The
      strongest filter for finding coordinates and counters under
      repeated same-direction input. Pair with `reverted` (using a
      reverse-action sequence) to fully confirm.

All addresses are linear (0..0x1FFFFF for the 2 MB virtual RAM). Output is
one candidate per line, sorted, with the byte values from each snapshot in
the order given. Pass --max N to cap output (default 200) and --width
{1,2,4} to group bytes as 8/16/32-bit values for readability.
"""
from __future__ import annotations

import argparse
import sys
from pathlib import Path


RAM_SIZE = 1 << 21  # 2 MB, matches SHIM_MEMORY_SIZE


def _load(path: str) -> bytes:
    data = Path(path).read_bytes()
    if len(data) != RAM_SIZE:
        raise SystemExit(
            f"state_discover: {path} is {len(data)} bytes, expected {RAM_SIZE}"
        )
    return data


def _format_value(snap: bytes, addr: int, width: int) -> str:
    chunk = snap[addr : addr + width]
    if width == 1:
        return f"{chunk[0]:02X}"
    if width == 2:
        v = chunk[0] | (chunk[1] << 8)
        return f"{v:04X}"
    if width == 4:
        v = chunk[0] | (chunk[1] << 8) | (chunk[2] << 16) | (chunk[3] << 24)
        return f"{v:08X}"
    raise ValueError(width)


def _addresses(width: int) -> range:
    return range(0, RAM_SIZE, width)


def _emit(candidates: list[int], snaps: list[bytes], width: int, limit: int) -> None:
    if not candidates:
        print("(no candidates)")
        return
    shown = candidates[:limit]
    for addr in shown:
        seg = (addr >> 4) & 0xFFFF
        off = addr & 0xF
        vals = " → ".join(_format_value(s, addr, width) for s in snaps)
        print(f"0x{addr:06X}  {seg:04X}:{off:04X}  {vals}")
    if len(candidates) > limit:
        print(f"... {len(candidates) - limit} more (use --max to widen)")
    print(f"# {len(candidates)} total candidate{'s' if len(candidates) != 1 else ''}")


def cmd_diff(args: argparse.Namespace) -> None:
    a = _load(args.snap_a)
    b = _load(args.snap_b)
    out: list[int] = []
    for addr in _addresses(args.width):
        if a[addr : addr + args.width] != b[addr : addr + args.width]:
            out.append(addr)
    _emit(out, [a, b], args.width, args.max)


def cmd_changed(args: argparse.Namespace) -> None:
    snaps = [_load(p) for p in args.snaps]
    if len(snaps) < 2:
        raise SystemExit("changed needs >=2 snapshots")
    out: list[int] = []
    for addr in _addresses(args.width):
        if all(
            snaps[i][addr : addr + args.width]
            != snaps[i + 1][addr : addr + args.width]
            for i in range(len(snaps) - 1)
        ):
            out.append(addr)
    _emit(out, snaps, args.width, args.max)


def cmd_reverted(args: argparse.Namespace) -> None:
    snaps = [_load(p) for p in args.snaps]
    if len(snaps) < 3:
        raise SystemExit("reverted needs >=3 snapshots (baseline, after, back)")
    baseline = snaps[0]
    after = snaps[1]
    laters = snaps[2:]
    out: list[int] = []
    for addr in _addresses(args.width):
        b = baseline[addr : addr + args.width]
        a = after[addr : addr + args.width]
        if a == b:
            continue
        if any(later[addr : addr + args.width] == b for later in laters):
            out.append(addr)
    _emit(out, snaps, args.width, args.max)


def _value_at(snap: bytes, addr: int, width: int) -> int:
    chunk = snap[addr : addr + width]
    v = 0
    for i, b in enumerate(chunk):
        v |= b << (8 * i)
    return v


def cmd_monotonic(args: argparse.Namespace) -> None:
    snaps = [_load(p) for p in args.snaps]
    if len(snaps) < 3:
        raise SystemExit("monotonic needs >=3 snapshots")
    out: list[int] = []
    for addr in _addresses(args.width):
        vals = [_value_at(s, addr, args.width) for s in snaps]
        if all(vals[i] < vals[i + 1] for i in range(len(vals) - 1)):
            out.append(addr)
            continue
        if all(vals[i] > vals[i + 1] for i in range(len(vals) - 1)):
            out.append(addr)
    _emit(out, snaps, args.width, args.max)


def cmd_constant(args: argparse.Namespace) -> None:
    snaps = [_load(p) for p in args.snaps]
    if len(snaps) < 2:
        raise SystemExit("constant needs >=2 snapshots")
    first = snaps[0]
    out: list[int] = []
    for addr in _addresses(args.width):
        chunk = first[addr : addr + args.width]
        if all(s[addr : addr + args.width] == chunk for s in snaps[1:]):
            out.append(addr)
    _emit(out, snaps, args.width, args.max)


def main() -> None:
    parser = argparse.ArgumentParser(
        description=__doc__,
        formatter_class=argparse.RawDescriptionHelpFormatter,
    )
    parser.add_argument(
        "--max", type=int, default=200,
        help="cap rows printed (default 200)")
    parser.add_argument(
        "--width", type=int, choices=(1, 2, 4), default=1,
        help="group bytes as 8/16/32-bit values (default 1)")
    sub = parser.add_subparsers(dest="command", required=True)

    sp = sub.add_parser("diff", help="byte-level diff of two snapshots")
    sp.add_argument("snap_a")
    sp.add_argument("snap_b")

    sp = sub.add_parser("changed", help="addresses that differ across every step")
    sp.add_argument("snaps", nargs="+")

    sp = sub.add_parser("reverted", help="changed in S1 then returned to S0 in some later Sn")
    sp.add_argument("snaps", nargs="+")

    sp = sub.add_parser("constant", help="addresses identical across all snapshots")
    sp.add_argument("snaps", nargs="+")

    sp = sub.add_parser(
        "monotonic",
        help="value strictly increases or strictly decreases at every step")
    sp.add_argument("snaps", nargs="+")

    args = parser.parse_args()
    handlers = {
        "diff": cmd_diff, "changed": cmd_changed,
        "reverted": cmd_reverted, "constant": cmd_constant,
        "monotonic": cmd_monotonic,
    }
    handlers[args.command](args)


if __name__ == "__main__":
    # `head` and pipes close stdout early; don't backtrace.
    try:
        main()
    except BrokenPipeError:
        sys.stderr.close()

File: saisei/src/state_discover.rs
//! `saisei state-discover` — find game-state addresses by snapshot diffing.
//! . Reads full 2 MB RAM snapshots
//! (dumped by `saisei control snapshot`) and diffs them under several filters
//! to locate the linear addresses that back a piece of game state.
//!
//! Subcommands:
//! diff S0 S1                 byte-level diff of two snapshots
//! changed S0 S1 [S2 ...]     addresses that differ across every adjacent pair
//! reverted S0 S1 [S2 ...]    changed in S1 then returned to S0 in some later Sn
//! constant S0 S1 [S2 ...]    addresses identical across all snapshots
//! monotonic S0 S1 S2 [...]   value strictly increases/decreases at every step

use std::path::Path;
use std::process::exit;

const RAM_SIZE: usize = 1 << 21; // 2 MB, matches SHIM_MEMORY_SIZE

/// the reference SystemExit(msg): message to stderr, exit code 1.
fn die(msg: &str) -> ! {
    eprintln!("{msg}");
    exit(1)
}

fn usage_err(msg: &str) -> ! {
    eprintln!("usage: saisei state-discover [--max N] [--width {{1,2,4}}] {{diff,changed,reverted,constant,monotonic}} ...");
    eprintln!("state-discover: error: {msg}");
    exit(2)
}

fn load(path: &str) -> Vec<u8> {
    let data = std::fs::read(path)
        .unwrap_or_else(|e| die(&format!("state_discover: cannot read {path}: {e}")));
    if data.len() != RAM_SIZE {
        die(&format!(
            "state_discover: {path} is {} bytes, expected {RAM_SIZE}",
            data.len()
        ));
    }
    data
}

fn format_value(snap: &[u8], addr: usize, width: usize) -> String {
    match width {
        1 => format!("{:02X}", snap[addr]),
        2 => {
            let v = snap[addr] as u32 | (snap[addr + 1] as u32) << 8;
            format!("{v:04X}")
        }
        4 => {
            let v = snap[addr] as u32
                | (snap[addr + 1] as u32) << 8
                | (snap[addr + 2] as u32) << 16
                | (snap[addr + 3] as u32) << 24;
            format!("{v:08X}")
        }
        _ => unreachable!(),
    }
}

fn value_at(snap: &[u8], addr: usize, width: usize) -> u64 {
    let mut v: u64 = 0;
    for i in 0..width {
        v |= (snap[addr + i] as u64) << (8 * i);
    }
    v
}

fn emit(candidates: &[usize], snaps: &[&[u8]], width: usize, limit: usize) {
    let mut out = String::new();
    if candidates.is_empty() {
        out.push_str("(no candidates)\n");
        print_out(&out);
        return;
    }
    let shown = &candidates[..candidates.len().min(limit)];
    for &addr in shown {
        let seg = (addr >> 4) & 0xFFFF;
        let off = addr & 0xF;
        let vals: Vec<String> = snaps.iter().map(|s| format_value(s, addr, width)).collect();
        out.push_str(&format!(
            "0x{addr:06X}  {seg:04X}:{off:04X}  {}\n",
            vals.join(" \u{2192} ")
        ));
    }
    if candidates.len() > limit {
        out.push_str(&format!(
            "... {} more (use --max to widen)\n",
            candidates.len() - limit
        ));
    }
    out.push_str(&format!(
        "# {} total candidate{}\n",
        candidates.len(),
        if candidates.len() != 1 { "s" } else { "" }
    ));
    print_out(&out);
}

/// Write to stdout; on a broken pipe (`head`, closed pipe) exit quietly, like
/// the reference's BrokenPipeError guard.
fn print_out(s: &str) {
    use std::io::Write;
    let stdout = std::io::stdout();
    let mut h = stdout.lock();
    if h.write_all(s.as_bytes()).and_then(|_| h.flush()).is_err() {
        exit(0);
    }
}

fn addresses(width: usize) -> impl Iterator<Item = usize> {
    (0..RAM_SIZE).step_by(width)
}

fn cmd_diff(snap_a: &str, snap_b: &str, width: usize, max: usize) {
    let a = load(snap_a);
    let b = load(snap_b);
    let mut out: Vec<usize> = Vec::new();
    for addr in addresses(width) {
        if a[addr..addr + width] != b[addr..addr + width] {
            out.push(addr);
        }
    }
    emit(&out, &[&a, &b], width, max);
}

fn cmd_changed(paths: &[String], width: usize, max: usize) {
    let snaps: Vec<Vec<u8>> = paths.iter().map(|p| load(p)).collect();
    if snaps.len() < 2 {
        die("changed needs >=2 snapshots");
    }
    let mut out: Vec<usize> = Vec::new();
    for addr in addresses(width) {
        if (0..snaps.len() - 1)
            .all(|i| snaps[i][addr..addr + width] != snaps[i + 1][addr..addr + width])
        {
            out.push(addr);
        }
    }
    let refs: Vec<&[u8]> = snaps.iter().map(|s| s.as_slice()).collect();
    emit(&out, &refs, width, max);
}

fn cmd_reverted(paths: &[String], width: usize, max: usize) {
    let snaps: Vec<Vec<u8>> = paths.iter().map(|p| load(p)).collect();
    if snaps.len() < 3 {
        die("reverted needs >=3 snapshots (baseline, after, back)");
    }
    let baseline = &snaps[0];
    let after = &snaps[1];
    let laters = &snaps[2..];
    let mut out: Vec<usize> = Vec::new();
    for addr in addresses(width) {
        let b = &baseline[addr..addr + width];
        let a = &after[addr..addr + width];
        if a == b {
            continue;
        }
        if laters.iter().any(|later| &later[addr..addr + width] == b) {
            out.push(addr);
        }
    }
    let refs: Vec<&[u8]> = snaps.iter().map(|s| s.as_slice()).collect();
    emit(&out, &refs, width, max);
}

fn cmd_constant(paths: &[String], width: usize, max: usize) {
    let snaps: Vec<Vec<u8>> = paths.iter().map(|p| load(p)).collect();
    if snaps.len() < 2 {
        die("constant needs >=2 snapshots");
    }
    let first = &snaps[0];
    let mut out: Vec<usize> = Vec::new();
    for addr in addresses(width) {
        let chunk = &first[addr..addr + width];
        if snaps[1..].iter().all(|s| &s[addr..addr + width] == chunk) {
            out.push(addr);
        }
    }
    let refs: Vec<&[u8]> = snaps.iter().map(|s| s.as_slice()).collect();
    emit(&out, &refs, width, max);
}

fn cmd_monotonic(paths: &[String], width: usize, max: usize) {
    let snaps: Vec<Vec<u8>> = paths.iter().map(|p| load(p)).collect();
    if snaps.len() < 3 {
        die("monotonic needs >=3 snapshots");
    }
    let mut out: Vec<usize> = Vec::new();
    for addr in addresses(width) {
        let vals: Vec<u64> = snaps.iter().map(|s| value_at(s, addr, width)).collect();
        let strictly_inc = (0..vals.len() - 1).all(|i| vals[i] < vals[i + 1]);
        if strictly_inc {
            out.push(addr);
            continue;
        }
        let strictly_dec = (0..vals.len() - 1).all(|i| vals[i] > vals[i + 1]);
        if strictly_dec {
            out.push(addr);
        }
    }
    let refs: Vec<&[u8]> = snaps.iter().map(|s| s.as_slice()).collect();
    emit(&out, &refs, width, max);
}

pub fn main(_root: &Path, args: &[String]) -> ! {
    let mut max: usize = 200;
    let mut width: usize = 1;
    let mut command: Option<String> = None;
    let mut positionals: Vec<String> = Vec::new();

    // Globals (--max/--width) precede the subcommand, mirroring argparse where
    // they live on the top-level parser; everything after the command token is
    // a positional for that subcommand.
    let mut it = args.iter();
    while let Some(a) = it.next() {
        if command.is_none() {
            let s = a.as_str();
            if let Some(v) = s.strip_prefix("--max=") {
                max = parse_max(v);
            } else if s == "--max" {
                let v = it
                    .next()
                    .cloned()
                    .unwrap_or_else(|| usage_err("argument --max: expected one argument"));
                max = parse_max(&v);
            } else if let Some(v) = s.strip_prefix("--width=") {
                width = parse_width(v);
            } else if s == "--width" {
                let v = it
                    .next()
                    .cloned()
                    .unwrap_or_else(|| usage_err("argument --width: expected one argument"));
                width = parse_width(&v);
            } else if s.starts_with("--") {
                usage_err(&format!("unrecognized arguments: {s}"));
            } else {
                command = Some(s.to_string());
            }
        } else {
            positionals.push(a.clone());
        }
    }

    let command =
        command.unwrap_or_else(|| usage_err("the following arguments are required: command"));

    match command.as_str() {
        "diff" => {
            if positionals.len() != 2 {
                usage_err("diff takes exactly two snapshots: diff S0 S1");
            }
            cmd_diff(&positionals[0], &positionals[1], width, max);
        }
        "changed" => {
            require_snaps(&positionals);
            cmd_changed(&positionals, width, max);
        }
        "reverted" => {
            require_snaps(&positionals);
            cmd_reverted(&positionals, width, max);
        }
        "constant" => {
            require_snaps(&positionals);
            cmd_constant(&positionals, width, max);
        }
        "monotonic" => {
            require_snaps(&positionals);
            cmd_monotonic(&positionals, width, max);
        }
        other => usage_err(&format!(
            "argument command: invalid choice: '{other}' (choose from 'diff', 'changed', 'reverted', 'constant', 'monotonic')"
        )),
    }
    exit(0)
}

fn require_snaps(positionals: &[String]) {
    // argparse nargs="+" requires at least one; the >=2 / >=3 checks live in the
    // handlers to reproduce the exact per-command error strings.
    if positionals.is_empty() {
        usage_err("the following arguments are required: snaps");
    }
}

fn parse_max(v: &str) -> usize {
    v.parse()
        .unwrap_or_else(|_| usage_err(&format!("argument --max: invalid int value: '{v}'")))
}

fn parse_width(v: &str) -> usize {
    match v {
        "1" => 1,
        "2" => 2,
        "4" => 4,
        _ => usage_err(&format!(
            "argument --width: invalid choice: '{v}' (choose from 1, 2, 4)"
        )),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn snap(fill: &[(usize, u8)]) -> Vec<u8> {
        let mut v = vec![0u8; RAM_SIZE];
        for &(a, b) in fill {
            v[a] = b;
        }
        v
    }

    #[test]
    fn diff_byte_logic() {
        let a = snap(&[(0x10, 1), (0x20, 5)]);
        let b = snap(&[(0x10, 2), (0x20, 5)]);
        let mut out = Vec::new();
        for addr in addresses(1) {
            if a[addr..addr + 1] != b[addr..addr + 1] {
                out.push(addr);
            }
        }
        assert_eq!(out, vec![0x10]);
    }

    #[test]
    fn value_at_little_endian() {
        let mut s = vec![0u8; RAM_SIZE];
        s[0x100] = 0x34;
        s[0x101] = 0x12;
        assert_eq!(value_at(&s, 0x100, 2), 0x1234);
        assert_eq!(format_value(&s, 0x100, 2), "1234");
        assert_eq!(format_value(&s, 0x100, 1), "34");
    }

    #[test]
    fn monotonic_increasing_and_decreasing() {
        // width-2 value at 0x200 goes 1 -> 2 -> 3 (increasing).
        let mk = |v: u16| {
            let mut s = vec![0u8; RAM_SIZE];
            s[0x200] = (v & 0xFF) as u8;
            s[0x201] = (v >> 8) as u8;
            s
        };
        let snaps = [mk(1), mk(2), mk(3)];
        let vals: Vec<u64> = snaps.iter().map(|s| value_at(s, 0x200, 2)).collect();
        assert!((0..vals.len() - 1).all(|i| vals[i] < vals[i + 1]));
    }

    #[test]
    fn width_parsing() {
        assert_eq!(parse_width("1"), 1);
        assert_eq!(parse_width("2"), 2);
        assert_eq!(parse_width("4"), 4);
    }
}
